src/params.rs

```diff
@@ -707,20 +709,35 @@ impl<'a> UnitaryGenerator for ParamManager {
                            , (h_init, (1, 2*hidden_size))];      // + the initial hidden state
 
         // biases next
-        let biases = vec![(h_bias_init, (1, 2*hidden_size))
+        let biases = vec![(h_bias_init, (1, hidden_size))
                           , (o_bias_init, (1, output_size))];
 
         // activations
         let activations = vec![h_activation, o_activation];
         
-        // we won't minimize trough the permutation params so we store them in optional
-        let optionals = Some(vec![(permut_init, (hidden_size, 1))]);
-        
+       
         self.add::<T>(manager, device, "unitary"
                       , weights
                       , biases
                       , activations
                       , None
-                      , optionals);
-  }
+                      , None);
+
+        // permutation and permutation inverse
+        let dims = Dim4::new(&[hidden_size as u64, 1, 1, 1]);
+        let seq: Vec<usize> = (0..hidden_size).collect();
+        let mut permut: Vec<u32> = Vec::with_capacity(seq.len());
+        // elementwise cast
+        for e in &seq {
+            permut.push(*e as u32);
+        }
+        rand::thread_rng().shuffle(&mut permut);
+        let mut permut_inv: Vec<u32> = permut.clone();
```

Comment:
Having a vec in the middle of layer operations removes the entire point of using the GPU.